File: Stack/PreviousLargest.java
import java.util.*;
class PreviousLargest
{
    public static void main(String args[])
    {
        Problem(new int[]{
            5,15,10,8,6,12,9,18
        });
    }
    private static void Problem(int[] arr)
    {
        int n=arr.length;
        int[] dp=new int[n];
        Stack<Integer> stack=new Stack<Integer>();
        stack.push(arr[0]);
        dp[0]=-1;
        for(int i=1;i<n;i++)
        {
            while(!(stack.empty()) && arr[i]>=stack.peek())
            {
                stack.pop();
            }
            if(stack.empty())
            {
                dp[i]=-1;
            }
            else{
                dp[i]=stack.peek();
            }
            stack.push(arr[i]);
        }
        for(int i=0;i<n;i++)
            System.out.print(dp[i]+" ");
    }
}

File: Stack/Previouslargest.py
import numpy as np;
def CalculateSpan(arr):
    stack=[]
    stack.append(arr[0])
    s=np.zeros(shape=len(arr),dtype=int)
    s[0]=-1
    for i in range(1,len(arr)):
        while(len(stack)>0 and stack[-1]<=arr[i]):
            stack.pop()
        s[i]=-1 if(len(stack)<=0) else(stack[-1])
        stack.append(arr[i])
       # print(stack)
    print(s)
CalculateSpan([20,30,10,5,15])